### PR TITLE
Add builder for macOS 10.15

### PR DESCRIPTION
**Pull request by mojca**
Not tested yet ...

### CODE REVIEW

**Comment by mojca** [comment]
We are getting some pretty unusual unit test failures which we might want to fix:
```
Failed to append AtomicParsley as a dependency to youtube-dl. Not Found.
Failed to append ffmpeg as a dependency to youtube-dl. Not Found.
Failed to append python37 as a dependency to youtube-dl. Not Found.
Failed to append zip as a dependency to youtube-dl. Not Found.
Failed to append python37 as a dependency to youtube-dl. Not Found.
Failed to append python27 as a dependency to py27-yapf. Not Found.
Failed to append py27-setuptools as a dependency to py27-yapf. Not Found.
Failed to append yapf_select as a dependency to py27-yapf. Not Found.
Failed to append python35 as a dependency to py35-yapf. Not Found.
Failed to append py35-setuptools as a dependency to py35-yapf. Not Found.
Failed to append yapf_select as a dependency to py35-yapf. Not Found.
Failed to append python36 as a dependency to py36-yapf. Not Found.
Failed to append py36-setuptools as a dependency to py36-yapf. Not Found.
Failed to append yapf_select as a dependency to py36-yapf. Not Found.
Failed to append python37 as a dependency to py37-yapf. Not Found.
Failed to append py37-setuptools as a dependency to py37-yapf. Not Found.
Failed to append yapf_select as a dependency to py37-yapf. Not Found.
Failed to append python38 as a dependency to py38-yapf. Not Found.
Failed to append py38-setuptools as a dependency to py38-yapf. Not Found.
Failed to append yapf_select as a dependency to py38-yapf. Not Found.
F.................
======================================================================
FAIL: test_updates (ports.tests.test_update_portinfo.TestUpdatePortinfo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/macports/macports-webapp/app/ports/tests/test_update_portinfo.py", line 18, in test_updates
    self.assertEquals(Port.objects.count(), 14)
AssertionError: 15 != 14
```

**Comment by arjunsalyan** [comment]
This test depends on macports-ports repository. It takes two fixed commits and carries out updates. Probably the commits I had chosen have been modified, but they were from months ago. I haven’t been able to take a thorough look, I will do this as soon as I have access to my computer.

**Comment by ryandesign** [comment]
Our repositories are configured to disallow modifications to past commits to master or release branches.

**Comment by mojca** [comment]
What has been more likely to happen is that the ports themselves were modified to a sufficient extent (and we are using the portindex from the server rather than generating it ourselves based on the given commit) that the test no longer passes as a consequence. I would need to take a closer look anyway.

**Comment by arjunsalyan** [comment]
The reason is: a new subport has been added to the port `py-yapf`, and hence the count of updated ports has increased by 1.

I will change the test to not count the ports but look for the parent port only. But then in future if any one of the ports py-yapf or YouTube-dl is deleted then the test will again start failing.

**Comment by arjunsalyan** [comment]
I have fixed the issue of the failing test, and this PR looks good to merged.

**Comment by mojca** [comment]
I forced pushed to the branch, but the Travil doesn't want to restart the tests.

We may still want to see if we can further improve the tests, but I agree that it's important to do two things fast:
* Add the 10.15 builder
* Make sure that full history gets fetched for all new builders.

Maybe we should add an environmental variable to the deployed version of our app (just this one instance) as a flag that certain tasks only need to be done there and not on copies done by users for the sake of testing.

**Comment by arjunsalyan** [comment]
I believe we can merge this after increasing the number of builds to be fetched. It is currently 20, but we can increase it as the number of builds for 10.15 would be much higher, while we decide on getting the entire build history.

**Comment by umeshksingla** [comment]
Do you mean we change it to 100 or 200 or something and then change it back to 20? I am good at it and don't think we ll have any problems.

**Comment by arjunsalyan** [comment]
> Do you mean we change it to 100 or 200 or something and then change it back to 20? I am good at it and don't think we ll have any problems.

Yes, but even without that, the PR looks good to be merged.

**Comment by umeshksingla** [comment]
@arjunsalyan I guess we can ignore codecov report for a migration file? Or maybe you want to take a look.

**Comment by ryandesign** [comment]
> Do you mean we change it to 100 or 200 or something and then change it back to 20?

Why wouldn't you want to get the information from all of the > 4000 builds that have happened so far on the 10.15 builder?

**Comment by arjunsalyan** [comment]
> Why wouldn't you want to get the information from all of the > 4000 builds that have happened so far on the 10.15 builder?

We can make it 4000 or more. But this would mean 4000+ continuous requests to the server. If we are good with that, then we can make it 4000 and try to fill that gap some other way.

**Comment by ryandesign** [comment]
I'm not familiar with the webapp code any why there is a limit at all, but I think the point is that we want build status information for as many ports as we can, so by all means please retrieve all the available info from the build server.

**Comment by ryandesign** [comment]
For the 10.15 builder, I would expect that you would start getting build information at build 1 and continue until the latest build, and then get incremental information later.

If increasing whatever this limit is to 4000 will also affect the other builders—for example if that would retrieve older builds for the new 10.6-10.8 builders than you have thus far retrieved—then I hope that it will correctly deal with the fact that builds happen chronologically. We would not want information about an old build of a port to overwrite information about a newer build of that port that had been previously retrieved. Each of the new 10.6-10.8 builders has fewer than 10,000 builds so far, which is how far back we keep build information, so you could go back to the beginning on those as well. But don't discard existing build information for 10.6-10.8, since we are only rebuilding ports that used libstdc++; builds of non-C++ software remain as they were before the new builders came online.

**Comment by arjunsalyan** [comment]
> For the 10.15 builder, I would expect that you would start getting build information at build 1 and continue until the latest build, and then get incremental information later.

I am seeing that builds start at number 0? Or is it 1?
This makes it easier to fetch all the builds.

> If increasing whatever this limit is to 4000 will also affect the other builders—for example if that would retrieve older builds for the new 10.6-10.8 builders than you have thus far retrieved—then I hope that it will correctly deal with the fact that builds happen chronologically. We would not want information about an old build of a port to overwrite information about a newer build of that port that had been previously retrieved. Each of the new 10.6-10.8 builders has fewer than 10,000 builds so far, which is how far back we keep build information, so you could go back to the beginning on those as well. But don't discard existing build information for 10.6-10.8, since we are only rebuilding ports that used libstdc++; builds of non-C++ software remain as they were before the new builders came online.

Currently, this will not affect other builders. The app first checks if it has any previous builds for a builder, if it does have- then it fetches only the new builds since the last fetch, otherwise it simply fetches 20 last builds.
For 10.15, I am going to make changes such that instead of fetching the last 20 builds, we start at 0.

**Comment by ryandesign** [comment]
> I am seeing that builds start at number 0? Or is it 1?

Oh yes, you are right, [the first build is 0](https://build.macports.org/builders/ports-10.15_x86_64-builder/builds/0).

**Comment by mojca** [comment]
Still not sure if you don't need to enter last build -1 to fetch build 0 first.

We should probably still do something to ensure that fetching from 0 won't happen on test deployments.

File: app/ports/migrations/0012_add_10.15_worker.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+def add_new_builder(apps, schema_editor):
+    Builder = apps.get_model('ports', 'Builder')
+    Builder.objects.create(name='10.15_x86_64', display_name='10.15')
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('ports', '0011_populate_display_name.py'),
```
Comment (arjunsalyan) [inline app/ports/migrations/0012_add_10.15_worker.py:12]: This migration did not get applied. After some testing I found that adding `.py` to the last migration's name causes it to fail, so I believe it to be the most probable reason.

We can try to fix this by removing this `.py` but I am spectacle about it, as it did not work for my development setup and I had to make a new migration.

Comment (arjunsalyan) [reply app/ports/migrations/0012_add_10.15_worker.py:12]: @neverpanic must have received an email regarding the failure.

Comment (neverpanic) [reply app/ports/migrations/0012_add_10.15_worker.py:12]: Haven't noticed. Maybe I missed it, though.

**Pull request merged**